mcs/class/Mono.Posix/Mono.Unix/UnixMarshal.cs

```diff
@@ -325,27 +325,34 @@ public static IntPtr StringToHeap (string s, int index, int count, Encoding enco
 			if (encoding == null)
 				throw new ArgumentNullException ("encoding");
 
-			int min_byte_count = encoding.GetMaxByteCount(1);
-			char[] copy = s.ToCharArray (index, count);
-			byte[] marshal = new byte [encoding.GetByteCount (copy) + min_byte_count];
+			int null_terminator_count = encoding.GetMaxByteCount (1);
```

Comment:
*I* know why it was done like this, as that was deliberate. ;-)

The `\0` terminator ir *not* always a single byte. In particular, since `Encoding` can be ~anything, e.g. `Encoding.Unicode` -- UTF-16 -- then the `\0` terminator could in fact be *two* bytes, *or more*! `Encoding.UTF32` would require *four* bytes.

Hence the `encoding.GetMaxByteCount()`, so that we can write null-terminated data using *any* encoding.

This is tested in `UnixMarshalTest.TestUtf32PtrToString()` and `UnixMarshalTest.TestStringToHeap()`.